Jenkinsfile

```
@@ -52,48 +44,55 @@ pipeline {
             }
         }
         stage('Publish final version images in dopplerdock') {
             environment {
                 DOCKER_CREDENTIALS_ID = "dockerhub_dopplerdock"
                 DOCKER_IMAGE_NAME = "dopplerdock/doppler-webapp"
             }
-            when {
-                expression {
-                    return isVersionTag(readCurrentTag())
+            stages {
+                stage('Publish pre-release images from pull request') {
+                    when {
+                        changeRequest target: 'master'
+                    }
+                    steps {
+                        withDockerRegistry(credentialsId: "${DOCKER_CREDENTIALS_ID}", url: "") {
+                            sh 'sh publish-commit-image-to-dockerhub.sh qa ${DOCKER_IMAGE_NAME} ${GIT_COMMIT} ${TAG_NAME}'
+                        }
+                    }
                 }
-            }
-            steps {
-                withDockerRegistry(credentialsId: "${DOCKER_CREDENTIALS_ID}", url: "") {
-                    sh 'sh publish-commit-image-to-dockerhub.sh production ${DOCKER_IMAGE_NAME} ${GIT_COMMIT} ${TAG_NAME}'
-                    sh 'sh publish-commit-image-to-dockerhub.sh demo ${DOCKER_IMAGE_NAME} ${GIT_COMMIT} ${TAG_NAME}'
-                    sh 'sh publish-commit-image-to-dockerhub.sh int ${DOCKER_IMAGE_NAME} ${GIT_COMMIT} ${TAG_NAME}'
-                    sh 'sh publish-commit-image-to-dockerhub.sh qa ${DOCKER_IMAGE_NAME} ${GIT_COMMIT} ${TAG_NAME}'
+                stage('Publish pre-release images from master') {
+                    when {
+                        branch 'master'
+                    }
+                    steps {
+                        withDockerRegistry(credentialsId: "${DOCKER_CREDENTIALS_ID}", url: "") {
+                            sh 'sh publish-commit-image-to-dockerhub.sh qa ${DOCKER_IMAGE_NAME} ${GIT_COMMIT} ${TAG_NAME}'
+                        }
+                    }
+                }
+                stage('Publish pre-release images from INT') {
+                    when {
+                        branch 'INT'
+                    }
+                    steps {
+                        withDockerRegistry(credentialsId: "${DOCKER_CREDENTIALS_ID}", url: "") {
+                            sh 'sh publish-commit-image-to-dockerhub.sh int ${DOCKER_IMAGE_NAME} ${GIT_COMMIT} ${TAG_NAME}'
+                        }
+                    }
+                }
+                stage('Publish final version images') {
+                    when {
+                        expression {
+                            return isVersionTag(readCurrentTag())
+                        }
+                    }
+                    steps {
+                        withDockerRegistry(credentialsId: "${DOCKER_CREDENTIALS_ID}", url: "") {
+                            sh 'sh publish-commit-image-to-dockerhub.sh production ${DOCKER_IMAGE_NAME} ${GIT_COMMIT} ${TAG_NAME}'
+                        }
+                    }
```

Creo que `publish-commit-image-to-dockerhub.sh` asume que ya exista una imagen preparada con un tag específico (por convención) y publica eso.

Creo que se podría hacer algo así (aunque separado por pasos sería mejor):


```suggestion
                    steps {
                        withDockerRegistry(credentialsId: "${DOCKER_CREDENTIALS_ID}", url: "") {
                            sh '''docker build \
                               -t "local-doppler-webapp:production-commit-${GIT_COMMIT}" \\
                               --build-arg environment=production \\
                               --build-arg react_app_router=browser \\
                               --build-arg public_url="/" \\
                               --build-arg version=production-${TAG_NAME}+${GIT_COMMIT} \\
                               -f Dockerfile.swarm \\
                            '''
                            sh 'sh publish-commit-image-to-dockerhub.sh production ${DOCKER_IMAGE_NAME} ${GIT_COMMIT} ${TAG_NAME}'
                        }
                    }
```

Lo mismo pero con el build específico para cada entorno.